In [1]:
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

import seaborn as sns
sns.set_context('notebook', font_scale=1.5, rc={'lines.markeredgewidth': 2})

In [2]:
%load_ext autoreload
%autoreload 2

%matplotlib inline

In [3]:
from visual_behavior.data_access import loading as loading
from visual_behavior.data_access import utilities 

from allensdk.brain_observatory.behavior.behavior_project_cache import VisualBehaviorOphysProjectCache

### get experiment and cell tables

In [4]:
experiments_table = loading.get_platform_paper_experiment_table()
cells_table = loading.get_cell_table()

In [5]:
# make sure date string column is there
experiments_table.date

ophys_experiment_id
775614751     20181108
778644591     20181113
788490510     20181129
796106850     20181215
792815735     20181210
                ...   
1086910729    20210225
1087430659    20210301
1087825608    20210303
1088351118    20210305
1088916626    20210308
Name: date, Length: 1249, dtype: int64

### look at columns for filtering and make sure they make sense

In [6]:
df = experiments_table.copy()

In [7]:
df[df.ophys_container_id==df.ophys_container_id.unique()[30]][['session_type', 'date_of_acquisition', 'date', 
                                            'experience_level', 'n_relative_to_first_novel', 'first_novel', 
                                            'second_novel', 'second_novel_active', 
                                            'last_familiar', 'last_familiar_active', 'prior_exposures_to_image_set']]

session_type         date_of_acquisition  \
ophys_experiment_id                                                         
962473290                    OPHYS_1_images_B  2019-10-07 16:54:33.000000   
963394065                    OPHYS_1_images_B  2019-10-08 16:36:06.000000   
964955813            OPHYS_2_images_B_passive  2019-10-09 17:34:04.000000   
965231035                    OPHYS_3_images_B  2019-10-10 18:03:17.000000   
967007328                    OPHYS_4_images_A  2019-10-14 15:15:27.000000   
967877560            OPHYS_5_images_A_passive  2019-10-15 16:40:48.000000   
969807132                    OPHYS_6_images_A  2019-10-17 15:01:11.000000   

                         date experience_level  n_relative_to_first_novel  \
ophys_experiment_id                                                         
962473290            20191007         Familiar                       -4.0   
963394065            20191008         Familiar                       -3.0   
964955813            20191009         Familiar                       -2.0   
965231035            20191010         Familiar                       -1.0   
967007328            20191014          Novel 1                        0.0   
967877560            20191015         Novel >1                        1.0   
969807132            20191017         Novel >1                        2.0   

                     first_novel  second_novel  second_novel_active  \
ophys_experiment_id                                                   
962473290                  False         False                False   
963394065                  False         False                False   
964955813                  False         False                False   
965231035                  False         False                False   
967007328                   True         False                False   
967877560                  False          True                False   
969807132                  False         False                 True   

                     last_familiar  last_familiar_active  \
ophys_experiment_id                                        
962473290                    False                 False   
963394065                    False                 False   
964955813                    False                 False   
965231035                     True                  True   
967007328                    False                 False   
967877560                    False                 False   
969807132                    False                 False   

                     prior_exposures_to_image_set  
ophys_experiment_id                                
962473290                                    18.0  
963394065                                    19.0  
964955813                                    20.0  
965231035                                    21.0  
967007328                                     0.0  
967877560                                     1.0  
969807132                                     3.0

## evaluate different filtering steps on cells table

### limit to containers with 3 exp levels

In [8]:
df = utilities.limit_to_containers_with_all_experience_levels(cells_table)

In [9]:
df.groupby(['ophys_container_id', 'experience_level']).count().reset_index().groupby(['ophys_container_id']).count()

experience_level  cell_specimen_id  ophys_experiment_id  \
ophys_container_id                                                            
791352433                          3                 3                    3   
803517529                          3                 3                    3   
803589437                          3                 3                    3   
812858755                          3                 3                    3   
814796612                          3                 3                    3   
...                              ...               ...                  ...   
1079028126                         3                 3                    3   
1080284671                         3                 3                    3   
1081683701                         3                 3                    3   
1084230237                         3                 3                    3   
1084230251                         3                 3                    3   

                    equipment_name  full_genotype  mouse_id  reporter_line  \
ophys_container_id                                                           
791352433                        3              3         3              3   
803517529                        3              3         3              3   
803589437                        3              3         3              3   
812858755                        3              3         3              3   
814796612                        3              3         3              3   
...                            ...            ...       ...            ...   
1079028126                       3              3         3              3   
1080284671                       3              3         3              3   
1081683701                       3              3         3              3   
1084230237                       3              3         3              3   
1084230251                       3              3         3              3   

                    driver_line  sex  age_in_days  ...  area_depth  \
ophys_container_id                                 ...               
791352433                     3    3            3  ...           3   
803517529                     3    3            3  ...           3   
803589437                     3    3            3  ...           3   
812858755                     3    3            3  ...           3   
814796612                     3    3            3  ...           3   
...                         ...  ...          ...  ...         ...   
1079028126                    3    3            3  ...           3   
1080284671                    3    3            3  ...           3   
1081683701                    3    3            3  ...           3   
1084230237                    3    3            3  ...           3   
1084230251                    3    3            3  ...           3   

                    area_binned_depth  date  first_novel  \
ophys_container_id                                         
791352433                           3     3            3   
803517529                           3     3            3   
803589437                           3     3            3   
812858755                           3     3            3   
814796612                           3     3            3   
...                               ...   ...          ...   
1079028126                          3     3            3   
1080284671                          3     3            3   
1081683701                          3     3            3   
1084230237                          3     3            3   
1084230251                          3     3            3   

                    n_relative_to_first_novel  last_familiar  \
ophys_container_id                                             
791352433                                   3              3   
803517529                                   3              3   
803589437                    

In [10]:
df.groupby(['ophys_container_id', 'experience_level']).count().reset_index().groupby(['ophys_container_id']).count().experience_level.unique()

array([3], dtype=int64)

There are only 3 experience levels per container

#### count number of experiments and cells

In [11]:
utilities.count_mice_expts_containers_cells(df)

n_mouse_id  n_ophys_experiment_id  \
cell_type      experience_level                                      
Excitatory     Familiar                  36                    227   
               Novel 1                   36                     71   
               Novel >1                  36                    142   
Sst Inhibitory Familiar                  15                    115   
               Novel 1                   15                     37   
               Novel >1                  15                     85   
Vip Inhibitory Familiar                  17                    138   
               Novel 1                   17                     44   
               Novel >1                  17                     96   

                                 n_ophys_container_id  n_cell_specimen_id  
cell_type      experience_level                                            
Excitatory     Familiar                            71               15268  
               Novel 1                             71                9225  
               Novel >1                            71               11962  
Sst Inhibitory Familiar                            37                 598  
               Novel 1                             37                 414  
               Novel >1                            37                 548  
Vip Inhibitory Familiar                            44                1233  
               Novel 1                             44                 742  
               Novel >1                            44                1067

There are many more cells and experiments for Familiar and Novel >1 because each container can have multiple of each of those. There is only one Novel 1 session per container by definition.

### limit to last familiar and second novel active

This should make the numbers of experiments and cells for each container more similar

In [12]:
df = utilities.limit_to_last_familiar_second_novel_active(cells_table)

In [13]:
df.groupby(['ophys_container_id', 'experience_level']).count().reset_index().groupby(['ophys_container_id']).count()

experience_level  cell_specimen_id  ophys_experiment_id  \
ophys_container_id                                                            
791352433                          3                 3                    3   
803517529                          3                 3                    3   
803589437                          3                 3                    3   
812858755                          2                 2                    2   
814796612                          3                 3                    3   
...                              ...               ...                  ...   
1079028126                         3                 3                    3   
1080284671                         3                 3                    3   
1081683701                         3                 3                    3   
1084230237                         3                 3                    3   
1084230251                         3                 3                    3   

                    equipment_name  full_genotype  mouse_id  reporter_line  \
ophys_container_id                                                           
791352433                        3              3         3              3   
803517529                        3              3         3              3   
803589437                        3              3         3              3   
812858755                        2              2         2              2   
814796612                        3              3         3              3   
...                            ...            ...       ...            ...   
1079028126                       3              3         3              3   
1080284671                       3              3         3              3   
1081683701                       3              3         3              3   
1084230237                       3              3         3              3   
1084230251                       3              3         3              3   

                    driver_line  sex  age_in_days  ...  area_depth  \
ophys_container_id                                 ...               
791352433                     3    3            3  ...           3   
803517529                     3    3            3  ...           3   
803589437                     3    3            3  ...           3   
812858755                     2    2            2  ...           2   
814796612                     3    3            3  ...           3   
...                         ...  ...          ...  ...         ...   
1079028126                    3    3            3  ...           3   
1080284671                    3    3            3  ...           3   
1081683701                    3    3            3  ...           3   
1084230237                    3    3            3  ...           3   
1084230251                    3    3            3  ...           3   

                    area_binned_depth  date  first_novel  \
ophys_container_id                                         
791352433                           3     3            3   
803517529                           3     3            3   
803589437                           3     3            3   
812858755                           2     2            2   
814796612                           3     3            3   
...                               ...   ...          ...   
1079028126                          3     3            3   
1080284671                          3     3            3   
1081683701                          3     3            3   
1084230237                          3     3            3   
1084230251                          3     3            3   

                    n_relative_to_first_novel  last_familiar  \
ophys_container_id                                             
791352433                                   3              3   
803517529                                   3              3   
803589437                    

In [14]:
df.groupby(['ophys_container_id', 'experience_level']).count().reset_index().groupby(['ophys_container_id']).count().experience_level.unique()

array([3, 2], dtype=int64)

There are some containers with only 2 experience levels, which means they are missing a Novel 1 session because we already filtered to select for the last Familiar and second Novel session, so we need to filter again to ensure we only have containers with all 3

### limit to last familiar and second novel active, and limit to containers with all 3 experience levels

its important to limit to last familiar and second novel active first, so that limiting to 3 experience levels only considers the last familiar and second novel active sessions

In [15]:
df = utilities.limit_to_last_familiar_second_novel_active(cells_table) # important that this goes first
df = utilities.limit_to_containers_with_all_experience_levels(df)

In [16]:
df.groupby(['ophys_container_id', 'experience_level']).count().reset_index().groupby(['ophys_container_id']).count().experience_level.unique()

array([3], dtype=int64)

There are only 3 experience levels per container

There should only be 1 experiment per experience level

In [17]:
df.groupby(['ophys_container_id', 'experience_level', 'ophys_experiment_id']).count().reset_index().groupby(['ophys_container_id', 'experience_level']).count().ophys_experiment_id.unique()

array([1], dtype=int64)

#### count numbers of expts and cells

There should be the same number of experiments for each experience level, and a similar number of cells

In [18]:
utilities.count_mice_expts_containers_cells(df)

n_mouse_id  n_ophys_experiment_id  \
cell_type      experience_level                                      
Excitatory     Familiar                  34                     62   
               Novel 1                   34                     62   
               Novel >1                  34                     62   
Sst Inhibitory Familiar                  15                     30   
               Novel 1                   15                     30   
               Novel >1                  15                     30   
Vip Inhibitory Familiar                  17                     42   
               Novel 1                   17                     42   
               Novel >1                  17                     42   

                                 n_ophys_container_id  n_cell_specimen_id  
cell_type      experience_level                                            
Excitatory     Familiar                            62                7302  
               Novel 1                             62                7974  
               Novel >1                            62                7404  
Sst Inhibitory Familiar                            30                 321  
               Novel 1                             30                 337  
               Novel >1                            30                 320  
Vip Inhibitory Familiar                            42                 794  
               Novel 1                             42                 718  
               Novel >1                            42                 765

### limit to cells matched in all 3 experience levels, including any Familiar or Novel >1 session 

In [19]:
df = utilities.limit_to_cell_specimen_ids_matched_in_all_experience_levels(cells_table)

In [20]:
utilities.count_mice_expts_containers_cells(df)

n_mouse_id  n_ophys_experiment_id  \
cell_type      experience_level                                      
Excitatory     Familiar                  36                    226   
               Novel 1                   36                     71   
               Novel >1                  36                    142   
Sst Inhibitory Familiar                  15                    112   
               Novel 1                   15                     36   
               Novel >1                  15                     84   
Vip Inhibitory Familiar                  17                    137   
               Novel 1                   17                     44   
               Novel >1                  17                     96   

                                 n_ophys_container_id  n_cell_specimen_id  
cell_type      experience_level                                            
Excitatory     Familiar                            71                5695  
               Novel 1                             71                5695  
               Novel >1                            71                5695  
Sst Inhibitory Familiar                            36                 292  
               Novel 1                             36                 292  
               Novel >1                            36                 292  
Vip Inhibitory Familiar                            44                 540  
               Novel 1                             44                 540  
               Novel >1                            44                 540

There are the same number of cell_specimen_ids for each experience level, but different numbers of experiment_ids, because we considered any Familiar or Novel >1 as a candidate for a match

In [21]:
df.groupby(['ophys_container_id', 'experience_level']).count().reset_index().groupby(['ophys_container_id']).count().experience_level.unique()

array([3], dtype=int64)

There are always 3 experience levels per container

In [22]:
df.groupby(['ophys_container_id', 'experience_level', 'ophys_experiment_id']).count().reset_index().groupby(['ophys_container_id', 'experience_level']).count().ophys_experiment_id.unique()

array([3, 1, 2, 4, 6, 5], dtype=int64)

But can be more than one experiment per experience level

### limit to cells matched in all 3 experience levels, only considering last Familiar and second Novel active

In [23]:
df = utilities.limit_to_last_familiar_second_novel_active(cells_table) # important that this goes first
df = utilities.limit_to_cell_specimen_ids_matched_in_all_experience_levels(df)

In [24]:
utilities.count_mice_expts_containers_cells(df)

n_mouse_id  n_ophys_experiment_id  \
cell_type      experience_level                                      
Excitatory     Familiar                  34                     62   
               Novel 1                   34                     62   
               Novel >1                  34                     62   
Sst Inhibitory Familiar                  14                     28   
               Novel 1                   14                     28   
               Novel >1                  14                     28   
Vip Inhibitory Familiar                  17                     41   
               Novel 1                   17                     41   
               Novel >1                  17                     41   

                                 n_ophys_container_id  n_cell_specimen_id  
cell_type      experience_level                                            
Excitatory     Familiar                            62                3306  
               Novel 1                             62                3306  
               Novel >1                            62                3306  
Sst Inhibitory Familiar                            28                 200  
               Novel 1                             28                 200  
               Novel >1                            28                 200  
Vip Inhibitory Familiar                            41                 415  
               Novel 1                             41                 415  
               Novel >1                            41                 415

Now the number of cell_specimen_ids AND the number of experiment_ids are the same across all 3 experience levels, because we have limited to only the last Familiar and first Novel active sessions

In [25]:
df.groupby(['ophys_container_id', 'experience_level']).count().reset_index().groupby(['ophys_container_id']).count().experience_level.unique()

array([3], dtype=int64)

There are 3 experience levels per container

In [26]:
df.groupby(['ophys_container_id', 'experience_level', 'ophys_experiment_id']).count().reset_index().groupby(['ophys_container_id', 'experience_level']).count().ophys_experiment_id.unique()

array([1], dtype=int64)

And only one experiment per experience level per container

#### This is the most conservative set of experiments and cells - matched cells across experience levels, only considering the most recent Familiar and Novel >1 sessions

### What is the number of image set exposures when we limit to last familiar and second novel active?

In [27]:
df = utilities.limit_to_last_familiar_second_novel_active(cells_table) # important that this goes first
df = utilities.limit_to_cell_specimen_ids_matched_in_all_experience_levels(df)

In [28]:
df.groupby(['cell_type', 'experience_level', 'prior_exposures_to_image_set']).mean().reset_index().groupby(['cell_type', 'experience_level']).describe()[['prior_exposures_to_image_set']]

prior_exposures_to_image_set             \
                                                       count       mean   
cell_type      experience_level                                           
Excitatory     Familiar                                 22.0  28.636364   
               Novel 1                                   1.0   0.000000   
               Novel >1                                  4.0   2.500000   
Sst Inhibitory Familiar                                 13.0  30.923077   
               Novel 1                                   1.0   0.000000   
               Novel >1                                  3.0   2.000000   
Vip Inhibitory Familiar                                 15.0  31.400000   
               Novel 1                                   1.0   0.000000   
               Novel >1                                  4.0   3.000000   

                                                                            
                                       std   min    25%   50%    75%   max  
cell_type      experience_level                                             
Excitatory     Familiar          14.257982  10.0  17.00  27.5  35.25  59.0  
               Novel 1                 NaN   0.0   0.00   0.0   0.00   0.0  
               Novel >1           1.290994   1.0   1.75   2.5   3.25   4.0  
Sst Inhibitory Familiar          16.378144  10.0  18.00  29.0  42.00  68.0  
               Novel 1                 NaN   0.0   0.00   0.0   0.00   0.0  
               Novel >1           1.000000   1.0   1.50   2.0   2.50   3.0  
Vip Inhibitory Familiar          15.393876  11.0  19.00  26.0  46.50  53.0  
               Novel 1                 NaN   0.0   0.00   0.0   0.00   0.0  
               Novel >1           2.160247   1.0   1.75   2.5   3.75   6.0

* For Excitatory cells, the average exposure number for Novel >1 is 2.5, with a std of 1.3
* For Sst cells, the average exposure number for Novel >1 is 2 with a std of 1
* For Vip cells, the average exposure number for Novel >1 is 3 with a std of 2.16

#### What if we dont limit to active sessions? 

### What is the number of image set exposures when we limit to last familiar and second novel, including passive sessions?

In [29]:
df = utilities.limit_to_last_familiar_second_novel(cells_table) # important that this goes first
df = utilities.limit_to_cell_specimen_ids_matched_in_all_experience_levels(df)

In [30]:
df.groupby(['cell_type', 'experience_level', 'prior_exposures_to_image_set']).mean().reset_index().groupby(['cell_type', 'experience_level']).describe()[['prior_exposures_to_image_set']]

prior_exposures_to_image_set             \
                                                       count       mean   
cell_type      experience_level                                           
Excitatory     Familiar                                 24.0  27.791667   
               Novel 1                                   1.0   0.000000   
               Novel >1                                  3.0   2.333333   
Sst Inhibitory Familiar                                 15.0  32.666667   
               Novel 1                                   1.0   0.000000   
               Novel >1                                  2.0   1.500000   
Vip Inhibitory Familiar                                 16.0  30.437500   
               Novel 1                                   1.0   0.000000   
               Novel >1                                  3.0   2.000000   

                                                                            
                                       std   min    25%   50%    75%   max  
cell_type      experience_level                                             
Excitatory     Familiar          14.255370   9.0  15.75  26.0  34.50  59.0  
               Novel 1                 NaN   0.0   0.00   0.0   0.00   0.0  
               Novel >1           1.527525   1.0   1.50   2.0   3.00   4.0  
Sst Inhibitory Familiar          18.148855  11.0  18.50  29.0  43.50  68.0  
               Novel 1                 NaN   0.0   0.00   0.0   0.00   0.0  
               Novel >1           0.707107   1.0   1.25   1.5   1.75   2.0  
Vip Inhibitory Familiar          15.798602  11.0  17.50  24.5  45.25  54.0  
               Novel 1                 NaN   0.0   0.00   0.0   0.00   0.0  
               Novel >1           1.000000   1.0   1.50   2.0   2.50   3.0

* For Excitatory cells, the average exposure number for Novel >1 is 2.3, with a std of 1.5
* For Sst cells, the average exposure number for Novel >1 is 1.5 with a std of 0.7
* For Vip cells, the average exposure number for Novel >1 is 2 with a std of 1

### Quantify exposure numbers per experience level for full dataset

First limit to containers with all experience levels so we dont consider containers without a first Novel session

In [31]:
df = utilities.limit_to_containers_with_all_experience_levels(cells_table)

Note that there can be more than one experiment per experience level in this scenario

In [32]:
df.groupby(['cell_type', 'experience_level', 'prior_exposures_to_image_set']).mean().reset_index().groupby(['cell_type', 'experience_level']).describe()[['prior_exposures_to_image_set']]

prior_exposures_to_image_set             \
                                                       count       mean   
cell_type      experience_level                                           
Excitatory     Familiar                                 39.0  29.384615   
               Novel 1                                   1.0   0.000000   
               Novel >1                                  6.0   3.500000   
Sst Inhibitory Familiar                                 36.0  32.805556   
               Novel 1                                   1.0   0.000000   
               Novel >1                                  6.0   3.500000   
Vip Inhibitory Familiar                                 32.0  31.000000   
               Novel 1                                   1.0   0.000000   
               Novel >1                                  7.0   4.000000   

                                                                            
                                       std   min    25%   50%    75%   max  
cell_type      experience_level                                             
Excitatory     Familiar          14.854762   8.0  17.50  28.0  37.50  59.0  
               Novel 1                 NaN   0.0   0.00   0.0   0.00   0.0  
               Novel >1           1.870829   1.0   2.25   3.5   4.75   6.0  
Sst Inhibitory Familiar          18.102596   9.0  18.75  28.5  43.50  69.0  
               Novel 1                 NaN   0.0   0.00   0.0   0.00   0.0  
               Novel >1           1.870829   1.0   2.25   3.5   4.75   6.0  
Vip Inhibitory Familiar          15.046166  10.0  17.75  26.5  46.25  54.0  
               Novel 1                 NaN   0.0   0.00   0.0   0.00   0.0  
               Novel >1           2.160247   1.0   2.50   4.0   5.50   7.0

* For Excitatory cells, the average exposure number for Novel >1 is 3.5, with a std of 1.8
* For Sst cells, the average exposure number for Novel >1 is 3.5 with a std of 1.8
* For Vip cells, the average exposure number for Novel >1 is 4 with a std of 2.16

### Limit to containers with all 3 experience levels, where the Novel >1 session is the second exposure to novelty

In [33]:
df = utilities.limit_to_containers_with_all_experience_levels(cells_table)

In [34]:
df = utilities.limit_to_second_novel_exposure(df)

In [35]:
# # code moved to utilities, but this is what id does to limit to second exposure
# # drop novel >1 sessions that arent the second exposure (prior exposures = 1)
# indices = df[(df.experience_level == 'Novel >1') & (df.prior_exposures_to_image_set != 1)].index.values
# df = df.drop(labels=indices, axis=0)

In [36]:
# now limit to last familiar second novel and cells matched in all 3
df = utilities.limit_to_last_familiar_second_novel(df) # important that this goes first
df = utilities.limit_to_cell_specimen_ids_matched_in_all_experience_levels(df)

In [37]:
utilities.count_mice_expts_containers_cells(df)

n_mouse_id  n_ophys_experiment_id  \
cell_type      experience_level                                      
Excitatory     Familiar                  25                     56   
               Novel 1                   25                     56   
               Novel >1                  25                     56   
Sst Inhibitory Familiar                  13                     23   
               Novel 1                   13                     23   
               Novel >1                  13                     23   
Vip Inhibitory Familiar                  15                     36   
               Novel 1                   15                     36   
               Novel >1                  15                     36   

                                 n_ophys_container_id  n_cell_specimen_id  
cell_type      experience_level                                            
Excitatory     Familiar                            56                2925  
               Novel 1                             56                2925  
               Novel >1                            56                2925  
Sst Inhibitory Familiar                            23                 127  
               Novel 1                             23                 127  
               Novel >1                            23                 127  
Vip Inhibitory Familiar                            36                 378  
               Novel 1                             36                 378  
               Novel >1                            36                 378

Now we get way fewer cells. But we can still quantify our effects here to be sure that prior exposures does not have a major effect

#### Compare to when we allow Novel >1 to be any exposure number, including passive sessions

In [38]:
df = utilities.limit_to_last_familiar_second_novel(cells_table) # important that this goes first
df = utilities.limit_to_cell_specimen_ids_matched_in_all_experience_levels(df)

In [39]:
utilities.count_mice_expts_containers_cells(df)

n_mouse_id  n_ophys_experiment_id  \
cell_type      experience_level                                      
Excitatory     Familiar                  34                     67   
               Novel 1                   34                     67   
               Novel >1                  34                     67   
Sst Inhibitory Familiar                  15                     31   
               Novel 1                   15                     31   
               Novel >1                  15                     31   
Vip Inhibitory Familiar                  17                     43   
               Novel 1                   17                     43   
               Novel >1                  17                     43   

                                 n_ophys_container_id  n_cell_specimen_id  
cell_type      experience_level                                            
Excitatory     Familiar                            67                3672  
               Novel 1                             67                3672  
               Novel >1                            67                3672  
Sst Inhibitory Familiar                            31                 214  
               Novel 1                             31                 214  
               Novel >1                            31                 214  
Vip Inhibitory Familiar                            43                 447  
               Novel 1                             43                 447  
               Novel >1                            43                 447

#### What do we get if we require that Novel >1 is the second exposure and is an active session? 

In [40]:
df = utilities.limit_to_containers_with_all_experience_levels(cells_table)

In [41]:
# drop novel >1 sessions that arent the second exposure (prior exposures = 1)
indices = df[(df.experience_level == 'Novel >1') & (df.prior_exposures_to_image_set != 1)].index.values
df = df.drop(labels=indices, axis=0)

In [42]:
df = utilities.limit_to_last_familiar_second_novel_active(df) # important that this goes first
df = utilities.limit_to_cell_specimen_ids_matched_in_all_experience_levels(df)

In [43]:
utilities.count_mice_expts_containers_cells(df)

n_mouse_id  n_ophys_experiment_id  \
cell_type      experience_level                                      
Excitatory     Familiar                   7                     14   
               Novel 1                    7                     14   
               Novel >1                   7                     14   
Sst Inhibitory Familiar                   2                      2   
               Novel 1                    2                      2   
               Novel >1                   2                      2   
Vip Inhibitory Familiar                   6                     20   
               Novel 1                    6                     20   
               Novel >1                   6                     20   

                                 n_ophys_container_id  n_cell_specimen_id  
cell_type      experience_level                                            
Excitatory     Familiar                            14                1015  
               Novel 1                             14                1015  
               Novel >1                            14                1015  
Sst Inhibitory Familiar                             2                  11  
               Novel 1                              2                  11  
               Novel >1                             2                  11  
Vip Inhibitory Familiar                            20                 162  
               Novel 1                             20                 162  
               Novel >1                            20                 162

Surprisingly, there are a good number of Novel >1 experiments that are the second exposure and are active sessions (surprising because a passive session is typically supposed to come after a first novel session)

In [44]:
df.groupby(['cell_type', 'experience_level', 'session_type']).count()[['cell_specimen_id']]

cell_specimen_id
cell_type      experience_level session_type                      
Excitatory     Familiar         OPHYS_1_images_A                25
                                OPHYS_1_images_B                79
                                OPHYS_3_images_A               680
                                OPHYS_3_images_B               231
               Novel 1          OPHYS_4_images_A               310
                                OPHYS_4_images_B               705
               Novel >1         OPHYS_4_images_A               189
                                OPHYS_4_images_B               658
                                OPHYS_6_images_A               121
                                OPHYS_6_images_B                47
Sst Inhibitory Familiar         OPHYS_3_images_B                11
               Novel 1          OPHYS_4_images_A                11
               Novel >1         OPHYS_4_images_A                11
Vip Inhibitory Familiar         OPHYS_1_images_A                69
                                OPHYS_3_images_A                71
                                OPHYS_3_images_B                22
               Novel 1          OPHYS_4_images_A                22
                                OPHYS_4_images_B               140
               Novel >1         OPHYS_4_images_A                19
                                OPHYS_4_images_B                94
                                OPHYS_6_images_A                 3
                                OPHYS_6_images_B                46